src/ripple/rpc/handlers/DownloadShard.cpp

```diff
@@ -0,0 +1,320 @@
+        return next();
+    }
+
+    // The extracted root directory name must match the shard index
+    auto const shardIndex {dCtx->shards.begin()->first};
+    auto shardDir {file.parent_path() / std::to_string(shardIndex)};
+    if (!exists(shardDir))
+    {
+        JLOG(dCtx->j.error()) <<
+            "Shard id mismatches archive shard directory name";
+        return next();
+    }
+
+    // Import the shard into the shard database
+    if (dCtx->shardStore->importShard(shardIndex, shardDir, dCtx->validate))
+    {
+        JLOG(dCtx->j.debug()) <<
+            "Shard " << shardIndex << " downloaded and imported";
+    }
+    else
+    {
+        JLOG(dCtx->j.error()) <<
+            "Failed to import shard: " << shardDir.string();
+    }
+    return next();
+}
+
+/** RPC command that downloads and import shard archives.
+    {
+      shards: [{index: <integer>, url: <string>}]
+      validate: <bool> // optional, default is true
+    }
+
+    example:
+    {
+      "command": "download_shard",
+      "shards": [
+        {"index": 1, "url": "https://domain.com/1.tar.lz4"},
+        {"index": 5, "url": "https://domain.com/5.tar.lz4"}
+      ]
+    }
+*/
+Json::Value
+doDownloadShard(RPC::Context& context)
+{
+    using namespace boost::filesystem;
+    if (context.role != Role::ADMIN)
+        return rpcError(rpcNO_PERMISSION);
+
+    // The shard store must be configured
+    if (!context.app.getShardStore())
+        return rpcError(rpcNOT_ENABLED);
+
+    // Deny request if already downloading
+    if (context.app.getShardStore()->getNumPreShard())
+        return rpcError(rpcTOO_BUSY);
+
+    if (!context.params.isMember(jss::shards))
+        return RPC::missing_field_error(jss::shards);
+    if (!context.params[jss::shards].isArray() ||
+        context.params[jss::shards].size() == 0)
+    {
+        return RPC::make_error(
+            rpcINVALID_PARAMS,
+            RPC::expected_field_message(
+                jss::shards,
+                "an array"));
+    }
+
+    auto dCtx {std::make_shared<DownloadContext>()};
+    dCtx->shardStore = context.app.getShardStore();
+    dCtx->j = context.app.journal("DownloadShard");
+
+    // Validate shards
+    static const std::string ext {".tar.lz4"};
+    for (auto& it : context.params[jss::shards])
+    {
+        // Validate the index
+        if (!it.isMember(jss::index))
+            return RPC::missing_field_error(jss::index);
+        auto& jv {it[jss::index]};
+        if (!(jv.isUInt() || (jv.isInt() && jv.asInt() >= 0)))
+        {
+            return RPC::make_error(rpcINVALID_PARAMS,
+                RPC::expected_field_message(jss::index,
+                    "unsigned integer"));
+        }
+
+        // Validate the URL
+        if (!it.isMember(jss::url))
+            return RPC::missing_field_error(jss::url);
+        parsedURL url;
+        if (!parseUrl(url, it[jss::url].asString()) ||
+            url.scheme != "https" ||
```

Comment:
for scheme in particular, it might be nice to return "only https supported" because I don't know if that would be obvious to me as a caller otherwise.